File: solutions/zig/bob/1/bob.zig
const std = @import("std");

fn question(input: []const u8) bool {
    return std.mem.endsWith(u8, input, "?");
}

fn shout(input: []const u8) bool {
    for (input) |c| {
        if (std.ascii.isAlphabetic(c) and std.ascii.isLower(c))
            return false;
    }

    const upperCaseLetters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    return std.mem.indexOfAny(u8, input, upperCaseLetters) != null;
}

pub fn response(s: []const u8) []const u8 {
    const trimmed = std.mem.trim(u8, s, &std.ascii.whitespace);

    if (trimmed.len == 0) return "Fine. Be that way!";
    if (question(trimmed) and shout(trimmed)) return "Calm down, I know what I'm doing!";
    if (question(trimmed)) return "Sure.";
    if (shout(trimmed)) return "Whoa, chill out!";
    return "Whatever.";
}

File: solutions/haskell/bob/1/src/Bob.hs
module Bob (responseFor) where

import Control.Monad (liftM2)
import qualified Data.Char as C
import Data.List

responseFor :: String -> String
responseFor xs
  | null trimmed = "Fine. Be that way!"
  | isAllCaps trimmed && isQuestion trimmed = "Calm down, I know what I'm doing!"
  | isAllCaps trimmed = "Whoa, chill out!"
  | isQuestion trimmed = "Sure."
  | otherwise = "Whatever."
  where
    trimmed = allTrim xs
    allTrim = trimLeft . trimRight
    trimLeft = dropWhile C.isSpace
    trimRight = dropWhileEnd C.isSpace

    isQuestion = (==) '?' . last
    isAllCaps =
      liftM2 (&&) (all C.isUpper) (not . null)
        . filter C.isAlpha
